drv/riscv-plic-server/src/main.rs

```diff
@@ -0,0 +1,202 @@
+    fn enable_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<ExtIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                for irq in irqs.iter() {
+                    plic.unmask(0, irq.0.try_into().unwrap());
+                }
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(ExtIntCtrlError::IRQUnassigned)),
+        }
+    }
+
+    /// Completes the interrupt on the PLIC, allowing for a new one to come
+    /// through.
+    fn complete_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<ExtIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                for irq in irqs.iter() {
+                    plic.complete(0, irq.0.try_into().unwrap());
+                }
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(ExtIntCtrlError::IRQUnassigned)),
+        }
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl {
+    // The PLIC is only interested in interrupt notifications from the kernel
+    fn current_notification_mask(&self) -> u32 {
+        return PLIC_IRQ;
+    }
+
+    // An interrupt has come in.
+    // NOTE: Currently, the driver assumes that all interrupts come in on
+    //       Context 0.
+    fn handle_notification(&mut self, _bits: u32) {
+        let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+        loop {
+            let irq: u32 = match plic.claim(0) {
+                Some(irq) => core::primitive::u16::from(irq) as u32,
+                None => break,
+            };
+
+            // An error means an interrupt came in on a line that no task has
+            // ownership over.
+            let owner: (TaskId, u32) = match get_irq_owner(irq) {
+                Ok(owner) => owner,
+                Err(()) => continue,
+            };
+
+            let code = sys_post(owner.0, owner.1);
+
+            // The task that owns the line was restarted.
+            if code & FIRST_DEAD_CODE == FIRST_DEAD_CODE {
+                let new_task_id = TaskId::for_index_and_gen(
+                    owner.0 .0.into(),
+                    ((code & !FIRST_DEAD_CODE) as u8).into(),
+                );
+
+                // SAFETY: We already have the irq owner, so we know that this
+                // operation will succeed. No need to bother checking for errors.
+                unsafe {
```

Comment:
Is there really a benefit to doing this unsafe, rather than just a normal unwrap?

Reply:
The driver will have to check the result in case it's an error due to the irq not being assigned which is never the case as the irq would have been discarded in line 143. This prevents the need for that extra comparison.